smp/__init__.py

```diff
@@ -0,0 +1,577 @@
+            return [i for i in range(cpu_count())]
+
+def set_affinity(mask):
+    if sched_setaffinity:
+        sched_setaffinity(0, mask)
+    else:
+        if PsutilProcess:
+            p = PsutilProcess()
+            p.cpu_affinity(mask)
+        else:
+            if os.name == "posix":
+                omp_set_num_threads(len(mask))
+
+def set_proc_affinity(process_count, process_id):
+    if process_count == 1:
+        return
+
+    cpu_list = list(get_affinity())
+    cpu_count = len(cpu_list)
+
+    global oversubscription_factor
+    if cpu_count < oversubscription_factor:
+        oversubscription_factor = cpu_count
+
+    threads_per_process = oversubscription_factor
+    if cpu_count >= process_count:
+        threads_per_process = threads_per_process*int(round(float(cpu_count)
+                              /float(process_count)))
+
+    start_cpu = (process_id*threads_per_process) % cpu_count;
+    mask = [cpu_list[((start_cpu + i) % cpu_count)]
+            for i in range(threads_per_process)]
+    set_affinity(mask)
+
+    if os.name == "posix":
+        mkl_set_num_threads(threads_per_process)
+
+def affinity_worker27(inqueue, outqueue, initializer=None, initargs=(),
+                      maxtasks=None, process_count=1, process_id=0):
+    from multiprocessing.pool import worker
+    set_proc_affinity(process_count, process_id)
+    worker(inqueue, outqueue, initializer, initargs, maxtasks)
+
+class AffinityPool27(Pool):
+    def __init__(self, processes=None, initializer=None, initargs=(),
+                 maxtasksperchild=None):
+        if max_top_workers:
+            processes = int(max_top_workers)
+        Pool.__init__(self, processes, initializer, initargs,
+                      maxtasksperchild)
+
+    def _repopulate_pool(self):
+        """Bring the number of pool processes up to the specified number,
+        for use after reaping workers which have exited.
+        """
+        from multiprocessing.util import debug
+
+        base_id = len(self._pool);
+        for i in range(self._processes - len(self._pool)):
+            w = self.Process(target=affinity_worker27,
+                             args=(self._inqueue, self._outqueue,
+                                   self._initializer,
+                                   self._initargs, self._maxtasksperchild,
+                                   self._processes, base_id + i)
+                            )
+            self._pool.append(w)
+            w.name = w.name.replace('Process', 'PoolWorker')
+            w.daemon = True
+            w.start()
+            debug('added worker')
+
+def affinity_worker35(inqueue, outqueue, initializer=None, initargs=(),
+                      maxtasks=None, wrap_exception=False,
+                      process_count=1, process_id=0):
+    from multiprocessing.pool import worker
+    set_proc_affinity(process_count, process_id)
+    worker(inqueue, outqueue, initializer, initargs, maxtasks, wrap_exception)
```

Comment:
This function is almost identical with the 27 version. I see no reason for this code-duplication.